graspy/embed/casc.py

```diff
@@ -0,0 +1,481 @@
+        to handle sparse matrices that may have large slowly decaying spectrum.
+
+    check_lcc : bool , optional (defult = True)
+        Whether to check if input graph is connected. May result in non-optimal
+        results if the graph is unconnected. If True and input is unconnected,
+        a UserWarning is thrown. Not checking for connectedness may result in 
+        faster computation.
+
+    regularizer: int, float or None, optional (default=None)
+        Constant to be added to the diagonal of degree matrix. If None, average
+        node degree is added. If int or float, must be >= 0. Only used when 
+        ``form`` == 'R-DAD'.
+    rownorm : bool , optional (default=False)
+        if True .The clastering process will be based on row-normlized singular
+        vectors.
+    CCA : bool , optional (default=False)
+        If True uses CCA method. Recommended for large number of clusters.see[1]_.
+    Assortative : bool , optional (default=True)
+        If True uses Assortative CASC method. For SBM models Assortative means 
+        within-block probability is higher than trans-block probability.see[1]_.
+    
+  
+
+    See Also
+    --------
+    graspy.embed.selectSVD
+    graspy.embed.select_dimension
+    graspy.utils.to_laplace
+    https://github.com/bwlewis/irlbpy
+    sklearn.cluster.KMeans
+    
+
+    References
+    ----------
+        [1]N. Binkiewicz, J. T. Vogelstein, K. Rohe, Covariate-assisted spectral
+        clustering, Biometrika, Volume 104, Issue 2, June 2017, Pages 361–377, 
+        https://doi.org/10.1093/biomet/asx008
+        [2]Augmented Implicitly Restarted Lanczos Bidiagonalization Methods,
+        J. Baglama and L. Reichel, SIAM J. Sci. Comput. 2005
+    """
+
+    def __init__(
+        self,
+        CCA=False,
+        form="R-DAD",
+        n_components=None,
+        n_elbows=2,
+        algorithm="randomized",
+        n_iter=5,
+        check_lcc=True,
+        regularizer=1,
+        assortative=True,
+       
+        row_norm=False,
+        n_points=100
+    ):
+        super().__init__(
+            n_components=n_components,
+            n_elbows=n_elbows,
+            algorithm=algorithm,
+            n_iter=n_iter,
+            check_lcc=check_lcc,
+        )
+        self.form = form
+        self.regularizer = regularizer
+        self.assortative = assortative
+        self.row_norm=row_norm
+        self.n_points=n_points
+        self.CCA=CCA
+        
+
+    
+#    @timethis
+    def fit(self, graph, covariate_matrix,y=None):
+        """
+        Fit CASC model to input graph
+
+        
+
+        Parameters
+        ----------
+        graph : array_like or networkx.Graph
+            Input graph to embed. see graspy.utils.import_graph
+        covariate:array_like ,shape(n_verts,n_covariates)
+            Bernoulli Covariate Matrix of a graph. 
+
+        y : Ignored
+
+        Returns
+        -------
+        Returns a dict containing following keywords:

```

Comment:
May I just change this to fit_transform?
This is because this class include both embedding and clustering so it's hard to inherit the fit function from parent class(either BaseEmbedding or BaseClustering)

Reply:
This still needs to be addressed. Make a fit_transform for returning the embeddings. Make a fit_predict for returning clusters.

Reply:
Sorry, I think I'm confused here. Do I need to make 2 different method for returning the clusters and the embeddings? But if I do so there might be one of the following issues: One choise is following what we have in embedding, but this will result in someone need to do the calculation twice to get both clusters and embedding. The second way will be require to call fit before calling fit_transform or fit_predict. If you are ok with these I'll update the code very soon.

Reply:
You call fit in fit_transform and fit_predict. Do the clustering step in fit_predict.

Reply:
I think you have a misunderstanding about this algorithm. In this method the embedding results depends on the clustering result. It's not the same as traditional spectral embedding that embed first and clustering later. Please see tutorial and the paper for more detail on this. That's why I'm asking the question.

Reply:
Oh right. Yes just call fit and return clusters in fit_predict. fit_transform should still just return the embeddings

Reply:
Do you mean do it this way:
fit{...}
fit_transform{ fit(); return embedding}
fit_predict{ fit(); return clusters}
But this may result in double calculation the result.

Reply:
No it doesn't. in your dict, casc_svd is the embedding. cluster is the clustering result. save the results as class attributes. like self.cluster_ = cluster or something. look through other embeddings methods to see this

Reply:
Yes I know that. It's still the same problem.  
To make it more clear, if you follow the way other embedding did the code will looklike.
fit(self, input)
{....
self. clusters_=clusters;
self.embedding_= casc_svd;}
fit_transform( self, input)
{
self.fit(self, input)
return self.embedding_
}
fit_predict(self, input)
{
self.fit(self, input)
return self.clusters_
}
In this way if someone wants to get both clusters and embeddings they will  call:
embedding=casc.fit_transform(input);
clusters=casc.fit_predict(input);
So still fit() are called twice in both fit_transform and fit_predict.